rosbag2/include/rosbag2/writer.hpp

```diff
@@ -95,13 +97,27 @@ class ROSBAG2_PUBLIC Writer
    */
   virtual void write(std::shared_ptr<SerializedBagMessage> message);
 
+  /**
+   * Check if the current recording bagfile needs to be split and rolled over to new file.
+   *
+   * \return true if the bagfile should be split.
+   */
+  virtual bool should_split_bagfile() const;
+
 private:
   std::string uri_;
   std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory_;
   std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory_;
   std::shared_ptr<rosbag2_storage::storage_interfaces::ReadWriteInterface> storage_;
   std::unique_ptr<rosbag2_storage::MetadataIo> metadata_io_;
   std::unique_ptr<Converter> converter_;
+  uint64_t max_bagfile_size_;
+  std::vector<std::string> relative_file_paths_;
+  uint64_t message_count_;
+  std::unordered_map<std::string, TopicInformation> topics_;
+  rcutils_time_point_value_t start_time_;
+  rcutils_time_point_value_t end_time_;
```

Comment:
Thinking about this, could we not directly instantiate a BagMetaData struct here? Most of the things in the `generate_metadata_()` function could actually directly be offloaded.

Reply:
The reason I didn't go with that design is because I used an `unordered_map` instead of a `vector` for topics. If I stored an internal `BagMetadata`, I'd have two indexing of topics.

Reply:
Why would you need to index any of them? The metadata has all the index information you'd need. 
Agreed, the `std::vector<TopicInformation>` is not perfect for o(1) indexing, but that shouldn't be a big pain point IMO.

Reply:
I think I know how to resolve this. See the [below comment](https://github.com/ros2/rosbag2/pull/184#discussion_r338214426) on what I think I should do right now. I'll resolve this conversation once I implement something.